### PR TITLE
feat: Add option to skip errors during CSV parsing

### PR DESCRIPTION
**Pull request by sinspiral**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

Add `csv_skip_errors` option when parsing a CSV to skip the error and continue parsing the file. 

This is useful when importing a batch of CSV files and there are corrupted lines on them, like when these CSV files are from a 3rd party not under your control and contain faulty lines.

Right now, when using the `input.file` plugin, whenever there is a parsing error, it exits the file processing, leaving you with a big set of files still to process. This is frustrating as you may be able to fix that line, but if there are more errors, you need to do it all over again. When having big batches of files (hundreds) this becomes a very time-consuming task.

 With this option, we allow you to skip that faulty line and continue the process of the files.

### CODE REVIEW

**Comment by sinspiral** [comment]
!signed-cla

File: plugins/parsers/csv/parser.go [hunk 1]
```diff
@@ -169,6 +172,10 @@ func parseCSV(p *Parser, r io.Reader) ([]telegraf.Metric, error) {
 	for _, record := range table {
 		m, err := p.parseRecord(record)
 		if err != nil {
+			if p.SkipErrors {
+				log.Println("[parsers.csv] Malformed csv line", err.Error())
```
Comment (Hipska) [inline plugins/parsers/csv/parser.go:176]: Not sure this is a correct logging format..

**Review by Hipska — APPROVED** [review]

File: plugins/parsers/csv/parser.go [hunk 2]
```diff
@@ -169,6 +171,12 @@ func parseCSV(p *Parser, r io.Reader) ([]telegraf.Metric, error) {
 	for _, record := range table {
 		m, err := p.parseRecord(record)
 		if err != nil {
+			if p.SkipErrors {
+				if p.Log != nil {
+					p.Log.Infof("[parsers.csv] Malformed csv line: %s", err.Error())
```
Comment (Hipska) [inline plugins/parsers/csv/parser.go:176]: Stil not sure if this is a good format, could you paste an example of how the total error message in the log will look like?

At least it needs to be changed to 
```suggestion
					p.Log.Errorf("[parsers.csv] Malformed csv line: %w", err)
```
Could maybe be a warning as well, but I'm more concerned about the `[parsers.csv]` part..

**Review by Hipska — CHANGES_REQUESTED** [review]
@srebhan Can you have a look as well, I know you did some fix related to parsers and logs..

File: plugins/parsers/csv/parser.go (hunk 2 above)
Comment (srebhan) [inline plugins/parsers/csv/parser.go:176]: The logger should always be set-up correctly, if it isn't, then this is a bug. So there is no need to check for the logger to be nil and there is no need to prefix the error message as this is done by the logger
```suggestion
				p.Log.Infof("Malformed line: %v", err)
```
You might also think of making this a `warn`, but on the otherhand this also has the potential to spam the log file. Not exactly sure what to do here... Maybe only log the first malformed line per run?

**Review by srebhan — COMMENTED** [review]
@sinspiral nice PR, thanks for contributing. I have one comment regarding the logging and TBH I'm not so sure what the best approach is... On the one hand, we want to inform the user on malformed lines and that this data is skipped. On the other hand, it has the potential to spam the log-file e.g. in cases where a lot of lines are malformed and you cannot do anything about it.... So one approach would be to make the log-entry a `debug`, but then chances are high that users don't notice... If you have a good idea, please speak up.

File: plugins/parsers/csv/parser.go (hunk 2 above)
Comment (Hipska) [reply plugins/parsers/csv/parser.go:176]: Good point, I agree Debug would be better than Warning or Error.

```suggestion
				p.Log.Debugf("parsing error: %v", err)
```

File: plugins/parsers/csv/parser.go [hunk 3]
```diff
@@ -169,6 +171,12 @@ func parseCSV(p *Parser, r io.Reader) ([]telegraf.Metric, error) {
...
+				}
+				continue
+			}
 			return metrics, err
```
Comment (Hipska) [inline plugins/parsers/csv/parser.go:180]: For consistency with above error message:
```suggestion
			return metrics, fmt.Errorf("parsing error: %w", err)
```

File: plugins/parsers/csv/parser.go (hunk 2 above)
Comment (sinspiral) [reply plugins/parsers/csv/parser.go]: Sorry for the delay... I was traveling for the holidays season. I also agree with the Debug statement. In my case, if I really need to check missing data, then I would enable the debug flag on the agent. I have updated with all the changes suggested. Thanks for the quick replies 👍 

Here is how it looks the messages when debug enabled:
`2021-12-22T11:36:30Z D! [parsers.csv::file] Parsing error: parsing time "corrupted_line" as "02/01/06 03:04:05 PM": cannot parse "corrupted_line" as "02"`

**Review by Hipska — APPROVED** [review]
Looks good, please also see my other comment

**Review by srebhan — APPROVED** [review]
Looks good to me. Thanks for working on this @sinspiral!

**Review by powersj — APPROVED** [review]

**Pull request merged**